File: actions_cpp/src/move_robot_client.cpp
#include "rclcpp/rclcpp.hpp"
#include "rclcpp_action/rclcpp_action.hpp"
#include "my_robot_interfaces/action/move_robot.hpp"
#include "my_robot_interfaces/msg/cancel_robot_move.hpp"

using MoveRobot = my_robot_interfaces::action::MoveRobot;
using MoveRobotGoalHandle = rclcpp_action::ClientGoalHandle<MoveRobot>;


class MoveRobotClientNode : public rclcpp::Node 
{
public:
    MoveRobotClientNode() : Node("move_robot_client") 
    {
        this->cancel_subscriber_ = this->create_subscription<my_robot_interfaces::msg::CancelRobotMove>(
            "cancel_move",
            10,
            std::bind(&MoveRobotClientNode::callback_cancel_move_robot, this, std::placeholders::_1)
        );

        this->move_robot_client_ = rclcpp_action::create_client<MoveRobot>(
            this,
            "move_robot"
        );
    }

    void send_goal(int position, int velocity)
    {
        // Wait for the action server
        move_robot_client_->wait_for_action_server();

        // Create the goal
        auto goal = MoveRobot::Goal();
        goal.position = position;
        goal.velocity = velocity;

        // Add callbacks
        auto options = rclcpp_action::Client<MoveRobot>::SendGoalOptions();
        options.feedback_callback = 
            std::bind(&MoveRobotClientNode::goal_feedback_callback, this, std::placeholders::_1, std::placeholders::_2);
        options.goal_response_callback = 
            std::bind(&MoveRobotClientNode::goal_response_callback, this, std::placeholders::_1);
        options.result_callback = 
            std::bind(&MoveRobotClientNode::goal_result_callback, this, std::placeholders::_1);

        // Send the goal
        RCLCPP_INFO(this->get_logger(), "Sending a goal");
        move_robot_client_->async_send_goal(goal, options);
    }

private:

    // Callback for cancel subscription
    void callback_cancel_move_robot(const my_robot_interfaces::msg::CancelRobotMove::SharedPtr msg)
    {
        (void)msg;
        if(this->goal_handle_)
        {
            RCLCPP_INFO(this->get_logger(), "Sending cancel request");
            move_robot_client_->async_cancel_goal(this->goal_handle_);
            this->goal_handle_.reset();
        }
    }

    // Callback to know if the goal request was accepted or rejected
    void goal_response_callback(const MoveRobotGoalHandle::SharedPtr &goal_handle)
    {
        if(!goal_handle)
        {
            RCLCPP_INFO(this->get_logger(), "Goal got rejected");
        }
        else{
            this->goal_handle_ = goal_handle;
            RCLCPP_INFO(this->get_logger(), "Goal got accepted");
        }
    }

    // Callback to receive the result once the goal was done
    void goal_result_callback(const MoveRobotGoalHandle::WrappedResult &result)
    {
        auto status = result.code;

        if(status==rclcpp_action::ResultCode::SUCCEEDED)
        {
            RCLCPP_INFO(this->get_logger(), "Succeded");
        }
        else if(status==rclcpp_action::ResultCode::ABORTED)
        {
            RCLCPP_ERROR(this->get_logger(), "Aborted");
        }
        else if(status==rclcpp_action::ResultCode::CANCELED)
        {
            RCLCPP_WARN(this->get_logger(), "Canceled");
        }

        int position = result.result->position;
        std::string message = result.result->message;
        RCLCPP_INFO(this->get_logger(), "Result: position: %d, message: %s", position, message.c_str());
    }

    void goal_feedback_callback(const MoveRobotGoalHandle::SharedPtr &goal_handle,
        const std::shared_ptr<const MoveRobot::Feedback> &feedback)
    {
        (void)goal_handle;
        int current_position = feedback->current_position;
        RCLCPP_INFO(this->get_logger(), "Got feedback: %d", current_position);
    }

    rclcpp_action::Client<MoveRobot>::SharedPtr move_robot_client_;
    MoveRobotGoalHandle::SharedPtr goal_handle_;
    rclcpp::Subscription<my_robot_interfaces::msg::CancelRobotMove>::SharedPtr cancel_subscriber_;
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<MoveRobotClientNode>(); 
    node->send_goal(20, 3);
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: actions_py/actions_py/move_robot_client.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from rclpy.action import ActionClient
from rclpy.action.client import ClientGoalHandle, GoalStatus
from my_robot_interfaces.action import MoveRobot
from my_robot_interfaces.msg import CancelRobotMove

class MoveRobotClientNode(Node): 
    def __init__(self):
        super().__init__("move_robot_client")

        self.goal_handle_ : ClientGoalHandle = None

        # Create a cancel topic
        self.cancel_subscriber_ = self.create_subscription(CancelRobotMove, "cancel_move",self.cancel_callback, 10)

        self.move_robot_client_ = ActionClient(self, MoveRobot, "move_robot")

    def cancel_callback(self,msg):
        if self.goal_handle_ is not None:
            self.get_logger().info("Sending a cancel request")
            self.goal_handle_.cancel_goal_async()

    def send_goal(self, position, velocity):
        # Wait for the server
        self.move_robot_client_.wait_for_server()

        # Create the goal 
        goal = MoveRobot.Goal()
        goal.position = position
        goal.velocity = velocity

        # Send the goal
        self.get_logger().info("Sending the goal")
        self.move_robot_client_.send_goal_async(goal,
                                                feedback_callback=self.goal_feedback_callback).\
                                                    add_done_callback(self.goal_response_callback)

    def goal_result_callback(self, future):
        # Get result from action
        status = future.result().status
        result = future.result().result

        if(status==GoalStatus.STATUS_SUCCEEDED):
            self.get_logger().info("Success")
        elif(status==GoalStatus.STATUS_ABORTED):
            self.get_logger().error("Aborted")
        elif(status==GoalStatus.STATUS_CANCELED):
            self.get_logger().warn("Canceled")
        self.get_logger().info("Robot position: "+str(result.position)+" message: "+result.message)

    def goal_feedback_callback(self, feedback_msg):
        current_position = feedback_msg.feedback.current_position
        self.get_logger().info("Got feedback: "+str(current_position))

    def goal_response_callback(self, future):
        self.goal_handle_ = future.result()

        # Check if the response was accepted
        if self.goal_handle_.accepted:
            self.get_logger().info("Goal got accepted")
            self.goal_handle_.get_result_async().add_done_callback(self.goal_result_callback)
        else:
            self.get_logger().warn("Goal got rejected")


def main(args=None):
    rclpy.init(args=args)
    node = MoveRobotClientNode()
    node.send_goal(34, 1) 
    rclpy.spin(node)
    rclpy.shutdown()


if __name__=="__main__":
    main()